code/modules/hydroponics/biogenerator.dm

```diff
@@ -31,307 +43,315 @@
 	RefreshParts()
 
 /obj/machinery/biogenerator/Destroy()
-	QDEL_NULL(beaker)
+	QDEL_NULL(container)
 	QDEL_NULL(files)
+	QDEL_LIST(stored_plants)
 	return ..()
 
 /obj/machinery/biogenerator/ex_act(severity)
-	if(beaker)
-		beaker.ex_act(severity)
+	container?.ex_act(severity)
 	..()
 
 /obj/machinery/biogenerator/handle_atom_del(atom/A)
 	..()
-	if(A == beaker)
-		beaker = null
-		update_icon()
-		updateUsrDialog()
+	if(A != container)
+		return
+	container = null
+	update_icon()
+	SStgui.update_uis(src)
 
 /obj/machinery/biogenerator/RefreshParts()
-	var/E = 0
-	var/P = 0
-	var/max_storage = 40
+	var/effeciency_prev = efficiency
+	efficiency = 0
+	productivity = 0
+	max_storable_plants = BASE_MAX_STORABLE_PLANTS
+
 	for(var/obj/item/stock_parts/matter_bin/B in component_parts)
-		P += B.rating
-		max_storage = 40 * B.rating
+		productivity += B.rating
+		max_storable_plants *= B.rating
 	for(var/obj/item/stock_parts/manipulator/M in component_parts)
-		E += M.rating
-	efficiency = E
-	productivity = P
-	max_items = max_storage
+		efficiency += M.rating
 
-/obj/machinery/biogenerator/on_reagent_change()			//When the reagents change, change the icon as well.
-	update_icon()
+	if(effeciency_prev >= 0 && effeciency_prev != efficiency)
+		update_ui_product_list() // We have have a higher `efficiency` now, and need to re-calc the product costs.
+		SStgui.update_uis(src, TRUE)
```

Comment:
move `SStgui.update_uis(src, TRUE)` inside the `update_ui_product_list()` then, since you no longer need the user.
Also it's probably better to pass named argument here (i.e. `update_static_data = TRUE)`).
I'ts often easy to infer what would be the meaning of an argument from the context, when said argument is an object (`src` / `user` / `item`), or string, or (depending on the name of the proc) number, but with bools it becomes way way harder.